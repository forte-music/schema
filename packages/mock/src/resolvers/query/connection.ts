import { Connection } from '../../models';

export interface ConnectionArgs {
  after?: string;
  first?: number;
}

export const handleConnection = <InputType, NodeType>(
  keys: InputType[],
  getNode: (key: InputType) => NodeType,
  { first = 25, after }: ConnectionArgs
): Connection<NodeType> => {
  const lowerBound = after ? parseInt(after, 10) + 1 : 0;
  const upperBound = first === -1 ? keys.length : lowerBound + first;

  const acceptedKeys = keys.slice(lowerBound, upperBound);

  return {
    count: keys.length,
    pageInfo: {
      hasNextPage: upperBound < keys.length,
    },
    edges: acceptedKeys.map((key, index) => ({
      cursor: (index + lowerBound).toString(),
      node: getNode(key),
    })),
  };
};
