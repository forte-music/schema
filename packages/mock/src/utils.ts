type Identifiable = {
  id: string;
};

export const mustGet = <K, V>(map: Map<K, V>, key: K): V => {
  const value = map.get(key);

  if (value === undefined) {
    throw new TypeError(`Schema invalid. Value not found in map. Key: ${key}`);
  }

  return value;
};

export const mustGetKeys = <T>(map: Map<string, T>, keys: string[]): T[] =>
  keys.map(key => mustGet(map, key));

// The getMap is needed because due to circular imports, the map may be
// undefined at import time. For example if song imports album's map and
// album import's songs map album's song map will be undefined because song
// hasn't exported anything yet.
export const arrayPropertyDescriptor = <T>(
  getMap: () => Map<string, T>,
  keys: string[]
) => ({
  get: () => mustGetKeys(getMap(), keys),
});

export const propertyDescriptor = <T>(
  getMap: () => Map<string, T>,
  key: string
) => ({
  get: () => mustGet(getMap(), key),
});

export const propertyDescriptorWithSet = <T>(
  getMap: () => Map<string, T>,
  key: string
) => ({
  get: () => mustGet(getMap(), key),
  set: (val: T): void => {
    getMap().set(key, val);
  },
});

export const makeMap = <T extends Identifiable>(list: T[]): Map<string, T> =>
  list.reduce((map, identifiable: T) => {
    addToMap(map, identifiable);
    return map;
  }, new Map());

export const addToMap = <T extends Identifiable>(
  map: Map<string, T>,
  item: T
) => {
  map.set(item.id, item);
};

export const now = () => Math.floor(Date.now() / 1000);

export const uuidForNum = (num: number): string =>
  num
    .toString(16)
    .toLowerCase()
    .padStart(32, '0');
