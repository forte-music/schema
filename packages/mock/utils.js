// @flow
type Identifiable = {
  id: string,
};

export const mustGet = <K: string, V>(map: Map<K, V>, key: K): V => {
  const value = map.get(key);

  if (value === undefined) {
    throw new TypeError(`Schema invalid. Value not found in map. Key: ${key}`);
  }

  return value;
};

export const mustGetKeys = <T>(map: Map<string, T>, keys: string[]): T[] =>
  keys.map(key => mustGet(map, key));

// The getMap is needed because due to circular imports, the map may be
// undefined at import time. For example if song imports album's map and
// album import's songs map album's song map will be undefined because song
// hasn't exported anything yet.
export const arrayPropertyDescriptor = <T>(
  getMap: () => Map<string, T>,
  keys: string[]
) => ({
  get: () => mustGetKeys(getMap(), keys),
});

export const propertyDescriptor = <T>(
  getMap: () => Map<string, T>,
  key: string
) => ({
  get: () => mustGet(getMap(), key),
});

export const propertyDescriptorWithSet = <T>(
  getMap: () => Map<string, T>,
  key: string
) => ({
  get: () => mustGet(getMap(), key),
  set: (val: T): void => {
    getMap().set(key, val);
  },
});

export const makeMap = <T: Identifiable>(list: T[]): Map<string, T> =>
  list.reduce((map, identifiable: T) => {
    map.set(identifiable.id, identifiable);
    return map;
  }, new Map());
