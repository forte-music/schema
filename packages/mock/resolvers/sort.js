// @flow
import type { Connection, UserStats } from '../models';
import type { ConnectionArgs } from './connection';
import { handleConnection } from './connection';

type SortBy =
  | 'RECENTLY_ADDED'
  | 'LEXICOGRAPHICALLY'
  | 'RECENTLY_PLAYED'
  | 'RELEVANCE';

type SortParams = {
  sortBy?: SortBy,
  reverse?: boolean,
  filter?: string,
};

type Sortable = {
  name: string,
  stats: UserStats,
  timeAdded: number,
};

type LinkedSortInfo<T> = {
  sortable: Sortable,
  original: T,
};

const compare = (a: any, b: any): number => {
  if (a > b) {
    return 1;
  }

  if (a === b) {
    return 0;
  }

  if (b > a) {
    return -1;
  }

  throw new TypeError(`Invariant a: ${a}, b: ${b}`);
};

// Returns how closely the record matches the query string.
const searchDistance = (record: string, query: string): number => {
  const normalizedQuery = query.toLowerCase().replace(/ /g, '');
  const normalizedRecord = record.toLowerCase();

  let strength = 0;
  let currentStrength = 0;
  let lastMatchIndex = 0;

  const khars = normalizedQuery.split('');
  for (const khar of khars) {
    const indexOfQueryCharacter = normalizedRecord.indexOf(
      khar,
      lastMatchIndex
    );
    if (indexOfQueryCharacter === -1) {
      // A character from the query isn't in the record. Sort to bottom.
      return -1;
    }

    if (lastMatchIndex === indexOfQueryCharacter - 1) {
      // continuation of last match
      currentStrength++;
    } else {
      // broken streak, reset streak strength
      currentStrength = 0;
    }

    // move last match index
    lastMatchIndex = indexOfQueryCharacter;

    // update strength
    strength += currentStrength;
  }

  return strength;
};

const reverseCompare = (a: any, b: any): number => reverse(a, b, compare);

const reverse = <T, R>(a: T, b: T, inner: (a: T, b: T) => R): R => inner(b, a);

const identity = <T, R>(a: T, b: T, inner: (a: T, b: T) => R): R => inner(a, b);

type Comparator<T> = (T, T) => number;

const sortWithReverse = <T>(
  reversed: boolean,
  inner: Comparator<T>
): Comparator<T> => {
  return (a: T, b: T): number => {
    const transform = reversed ? reverse : identity;
    return transform(a, b, inner);
  };
};

export type SortConnectionArgs = ConnectionArgs & {
  sort?: SortParams,
};

export const listItemsResolver = <T>(
  values: T[],
  { sort, ...connectionArgs }: SortConnectionArgs,
  process: (item: T) => Sortable
): Connection<T> => {
  const localHandleConnection = (items: T[]) =>
    handleConnection(items, value => value, connectionArgs);

  if (!sort) {
    // No sort was specified, use default sort.

    return localHandleConnection(values);
  }

  const { sortBy = 'LEXICOGRAPHICALLY', reverse = false, filter = '' } = sort;

  let sorter: Comparator<Sortable> = (a: Sortable, b: Sortable): number => 0;
  let filterFunc = (item: Sortable) => searchDistance(item.name, filter) !== -1;

  switch (sortBy) {
    case 'RECENTLY_ADDED':
      sorter = (a: Sortable, b: Sortable) =>
        reverseCompare(a.timeAdded, b.timeAdded);
      break;

    case 'LEXICOGRAPHICALLY':
      sorter = (a: Sortable, b: Sortable) => compare(a.name, b.name);
      break;

    case 'RECENTLY_PLAYED':
      sorter = (a: Sortable, b: Sortable) =>
        reverseCompare(a.stats.lastPlayed || 0, b.stats.lastPlayed || 0);
      break;

    case 'RELEVANCE':
      if (!filter) {
        throw new TypeError('Used relevance sort without filter.');
      }

      sorter = (a: Sortable, b: Sortable) =>
        reverseCompare(
          searchDistance(a.name, filter),
          searchDistance(b.name, filter)
        );

      break;

    default:
      throw new TypeError(`Unknown Sort ${sortBy}`);
  }

  const mappedValues: LinkedSortInfo<T>[] = values.map(item => ({
    original: item,
    sortable: process(item),
  }));

  const filteredValues = mappedValues.filter(({ original: T, sortable }) =>
    filterFunc(sortable)
  );

  const sortedValues = filteredValues.slice();
  sortedValues.sort(
    sortWithReverse(
      reverse,
      (
        { sortable: a }: LinkedSortInfo<T>,
        { sortable: b }: LinkedSortInfo<T>
      ) => sorter(a, b)
    )
  );

  return localHandleConnection(sortedValues.map(({ original }) => original));
};

export const itemsResolver = <T: Sortable>(map: Map<string, T>) => (
  _: void,
  args: SortConnectionArgs
): Connection<T> => {
  const values = Array.from(map.values());
  return listItemsResolver(values, args, item => item);
};
