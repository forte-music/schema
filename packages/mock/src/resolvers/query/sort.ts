import { Connection } from '../../models';
import { ConnectionArgs, handleConnection } from './connection';

type SortBy = 'RECENTLY_ADDED' | 'LEXICOGRAPHICALLY' | 'RECENTLY_PLAYED';

interface SortParams {
  sortBy?: SortBy;
  reverse?: boolean;
  filter?: string;
}

interface Sortable {
  name: string;
  lastPlayed?: number;
  timeAdded: number;
}

interface LinkedSortInfo<T> {
  sortable: Sortable;
  original: T;
}

interface Comparator<T> {
  (a: T, b: T): number;
}

export const compare = <T>(a: T, b: T): number => {
  if (a > b) {
    return 1;
  }

  if (a === b) {
    return 0;
  }

  if (b > a) {
    return -1;
  }

  throw new TypeError(`Invariant a: ${a}, b: ${b}`);
};

export const reverseCompare = <T>(a: T, b: T): number => reverse(a, b, compare);

const reverse = <T>(a: T, b: T, inner: Comparator<T>): number => inner(b, a);

const identity = <T>(a: T, b: T, inner: Comparator<T>): number => inner(a, b);

// Returns how closely the record matches the query string.
const searchDistance = (record: string, query: string): number => {
  const normalizedQuery = query.toLowerCase().replace(/ /g, '');
  const normalizedRecord = record.toLowerCase();

  let strength = 0;
  let currentStrength = 0;
  let lastMatchIndex = 0;

  const khars = normalizedQuery.split('');
  for (const khar of khars) {
    const indexOfQueryCharacter = normalizedRecord.indexOf(
      khar,
      lastMatchIndex
    );
    if (indexOfQueryCharacter === -1) {
      // A character from the query isn't in the record. Sort to bottom.
      return -1;
    }

    if (lastMatchIndex === indexOfQueryCharacter - 1) {
      // continuation of last match
      currentStrength++;
    } else {
      // broken streak, reset streak strength
      currentStrength = 0;
    }

    // move last match index
    lastMatchIndex = indexOfQueryCharacter;

    // update strength
    strength += currentStrength;
  }

  return strength;
};

const sortWithReverse = <T>(
  reversed: boolean,
  inner: Comparator<T>
): Comparator<T> => {
  return (a: T, b: T): number => {
    const transform = reversed ? reverse : identity;
    return transform(a, b, inner);
  };
};

export type SortConnectionArgs = ConnectionArgs & {
  sort?: SortParams;
};

export const listItemsResolver = <T>(
  values: T[],
  { sort, ...connectionArgs }: SortConnectionArgs,
  process: (item: T) => Sortable
): Connection<T> => {
  const localHandleConnection = (items: T[]) =>
    handleConnection(items, value => value, connectionArgs);

  if (!sort) {
    // No sort was specified, use default sort.

    return localHandleConnection(values);
  }

  const { sortBy = 'LEXICOGRAPHICALLY', reverse = false, filter = '' } = sort;

  let sorter: Comparator<Sortable> = () => 0;
  let filterFunc = (item: Sortable) => searchDistance(item.name, filter) !== -1;

  switch (sortBy) {
    case 'RECENTLY_ADDED':
      sorter = (a: Sortable, b: Sortable) =>
        reverseCompare(a.timeAdded, b.timeAdded);
      break;

    case 'LEXICOGRAPHICALLY':
      sorter = (a: Sortable, b: Sortable) => compare(a.name, b.name);
      break;

    case 'RECENTLY_PLAYED':
      sorter = (a: Sortable, b: Sortable) =>
        reverseCompare(a.lastPlayed || 0, b.lastPlayed || 0);
      break;

    default:
      throw new TypeError(`Unknown Sort ${sortBy}`);
  }

  const mappedValues: LinkedSortInfo<T>[] = values.map(item => ({
    original: item,
    sortable: process(item),
  }));

  const filteredValues = mappedValues.filter(({ sortable }) =>
    filterFunc(sortable)
  );

  const sortedValues = filteredValues.slice();
  sortedValues.sort(
    sortWithReverse(
      reverse,
      (
        { sortable: a }: LinkedSortInfo<T>,
        { sortable: b }: LinkedSortInfo<T>
      ) => sorter(a, b)
    )
  );

  return localHandleConnection(sortedValues.map(({ original }) => original));
};

export const itemsResolver = <T extends Sortable>(map: Map<string, T>) => (
  _: void,
  args: SortConnectionArgs
): Connection<T> => {
  const values = Array.from(map.values());
  return listItemsResolver(values, args, item => item);
};
