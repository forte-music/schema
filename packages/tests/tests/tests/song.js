import gql from 'graphql-tag';
import { print } from 'graphql/language/printer';

import { testConnection } from '../connection';
import { testSort } from '../sort';
import SongFields from './fragments/SongFields.graphql';
import AlbumFields from './fragments/AlbumFields.graphql';
import ArtistFields from './fragments/ArtistFields.graphql';

import client from '../client';
import { uuidForNum } from '../utils';

const variables = { songId: uuidForNum(2) };
it('should get a song by id', async () => {
  const query = gql`
    query($songId: ID!) {
      song(id: $songId) {
        ...SongFields

        album {
          ...AlbumFields
        }

        artists {
          ...ArtistFields
        }
      }
    }

    ${SongFields}
    ${AlbumFields}
    ${ArtistFields}
  `;

  const data = await client.request(print(query), variables);

  expect(data).toMatchSnapshot();
});

testConnection('songs', async ({ first, after }) => {
  const query = gql`
    query($first: Int, $after: String) {
      songs(first: $first, after: $after) {
        count
        pageInfo {
          hasNextPage
        }

        edges {
          cursor

          node {
            id
          }
        }
      }
    }
  `;

  const { songs } = await client.request(print(query), {
    first,
    after,
  });

  return songs;
});

testSort('songs', async ({ sortBy, reverse }) => {
  const query = gql`
    query($sortBy: SortBy!, $reverse: Boolean!) {
      songs(first: -1, sort: { sortBy: $sortBy, reverse: $reverse }) {
        edges {
          node {
            id
            name
            timeAdded
            lastPlayed
          }
        }
      }
    }
  `;

  const {
    songs: { edges },
  } = await client.request(print(query), {
    sortBy,
    reverse,
  });

  return edges.map(({ node }) => node);
});
