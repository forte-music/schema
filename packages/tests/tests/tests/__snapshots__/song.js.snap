// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should get a song by id 1`] = `
Object {
  "song": Object {
    "album": Object {
      "duration": 211,
      "id": "00000000000000000000000000000002",
      "lastPlayed": null,
      "name": "I'm That (Remix)",
      "releaseYear": 2015,
    },
    "artists": Array [
      Object {
        "id": "00000000000000000000000000000002",
        "lastPlayed": null,
        "name": "R. City",
      },
      Object {
        "id": "00000000000000000000000000000015",
        "lastPlayed": null,
        "name": "Beenie Man",
      },
      Object {
        "id": "00000000000000000000000000000016",
        "lastPlayed": null,
        "name": "Azealia Banks",
      },
    ],
    "diskNumber": 1,
    "id": "00000000000000000000000000000002",
    "lastPlayed": null,
    "liked": false,
    "name": "I'm That (Remix)",
    "playCount": 0,
    "trackNumber": 1,
  },
}
`;
